service/history/queues/slice.go

```diff
@@ -183,56 +195,88 @@ func (s *SliceImpl) splitExecutables(
 	return thisExecutable, thatExecutables
 }
 
-func (s *SliceImpl) CanMergeByRange(slice Slice) bool {
-	s.validateNotDestroyed()
+func (s *SliceImpl) CanMergeWithSlice(slice Slice) bool {
+	s.stateSanityCheck()
 
-	return s != slice && s.scope.CanMergeByRange(slice.Scope())
+	return s != slice && s.scope.Range.CanMerge(slice.Scope().Range)
 }
 
-func (s *SliceImpl) MergeByRange(slice Slice) Slice {
-	if !s.CanMergeByRange(slice) {
-		panic(fmt.Sprintf("Unable to merge queue slice having scope %v with slice having scope %v by range", s.scope, slice.Scope()))
+func (s *SliceImpl) MergeWithSlice(slice Slice) []Slice {
+	if s.scope.Range.InclusiveMin.CompareTo(slice.Scope().Range.InclusiveMin) > 0 {
+		return slice.MergeWithSlice(s)
+	}
+
+	if !s.CanMergeWithSlice(slice) {
+		panic(fmt.Sprintf("Unable to merge queue slice having scope %v with slice having scope %v", s.scope, slice.Scope()))
 	}
 
 	incomingSlice, ok := slice.(*SliceImpl)
 	if !ok {
 		panic(fmt.Sprintf("Unable to merge queue slice of type %T with type %T", s, slice))
 	}
 
-	s.destroy()
-	incomingSlice.destroy()
-	return &SliceImpl{
-		executableInitializer:  s.executableInitializer,
-		scope:                  s.scope.MergeByRange(incomingSlice.scope),
-		outstandingExecutables: s.mergeExecutables(incomingSlice),
-		iterators:              s.mergeIterators(incomingSlice),
+	if s.scope.CanMergeByRange(incomingSlice.scope) {
+		return []Slice{s.mergeByRange(incomingSlice)}
 	}
-}
-
-func (s *SliceImpl) CanMergeByPredicate(slice Slice) bool {
-	s.validateNotDestroyed()
 
-	return s != slice && s.scope.CanMergeByPredicate(slice.Scope())
-}
+	mergedSlices := make([]Slice, 0, 3)
+	currentLeftSlice, currentRightSlice := s.splitByRange(incomingSlice.Scope().Range.InclusiveMin)
+	if !currentLeftSlice.scope.IsEmpty() {
+		mergedSlices = append(mergedSlices, currentLeftSlice)
+	}
 
-func (s *SliceImpl) MergeByPredicate(slice Slice) Slice {
-	if !s.CanMergeByPredicate(slice) {
-		panic(fmt.Sprintf("Unable to merge queue slice having scope %v with slice having scope %v by predicate", s.scope, slice.Scope()))
+	if currentRightMax := currentRightSlice.Scope().Range.ExclusiveMax; incomingSlice.CanSplitByRange(currentRightMax) {
+		leftIncomingSlice, rightIncomingSlice := incomingSlice.splitByRange(currentRightMax)
+		mergedMidSlice := currentRightSlice.mergeByPredicate(leftIncomingSlice)
+		if !mergedMidSlice.scope.IsEmpty() {
+			mergedSlices = append(mergedSlices, mergedMidSlice)
+		}
+		if !rightIncomingSlice.scope.IsEmpty() {
+			mergedSlices = append(mergedSlices, rightIncomingSlice)
+		}
+	} else {
+		currentMidSlice, currentRightSlice := currentRightSlice.splitByRange(incomingSlice.Scope().Range.ExclusiveMax)
+		mergedMidSlice := currentMidSlice.mergeByPredicate(incomingSlice)
+		if !mergedMidSlice.scope.IsEmpty() {
```

Comment:
add an assertion that `currentMidSlice == incomingSlice`?

Reply:
The assertions of ranges are equal is done by `scope.MergeByPredicate` inside `mergeByPredicate()`